File: packages/qf-scaffolding/src/tools/project_files.ts
/**
 * Project file tools for qf-scaffolding MCP
 *
 * General-purpose file read/write within project directory.
 * Allows Claude to access ANY file within the project, not just
 * specific folders like materials/ or questions/.
 *
 * Security: Only allows access within project_path (no path traversal)
 */

import { z } from "zod";
import { readFile, writeFile, mkdir, stat } from "fs/promises";
import { join, dirname, normalize, relative, isAbsolute } from "path";
import { existsSync } from "fs";
import { logEvent, logError } from "../utils/logger.js";

// ============================================
// READ PROJECT FILE
// ============================================

export const readProjectFileSchema = z.object({
  project_path: z.string(),
  relative_path: z.string(), // Path relative to project_path, e.g. "output/questions.md"
});

export type ReadProjectFileInput = z.infer<typeof readProjectFileSchema>;

export interface ReadProjectFileResult {
  success: boolean;
  file_path?: string;
  relative_path?: string;
  content?: string;
  size_bytes?: number;
  error?: string;
}

/**
 * Validate that the resolved path is within project_path (prevent path traversal)
 */
function isPathWithinProject(projectPath: string, targetPath: string): boolean {
  const normalizedProject = normalize(projectPath);
  const normalizedTarget = normalize(targetPath);
  const relativePath = relative(normalizedProject, normalizedTarget);

  // If relative path starts with ".." or is absolute, it's outside project
  return (
    !relativePath.startsWith("..") &&
    !isAbsolute(relativePath) &&
    relativePath !== ""
  );
}

/**
 * Read any file within the project directory
 */
export async function readProjectFile(
  input: ReadProjectFileInput
): Promise<ReadProjectFileResult> {
  const { project_path, relative_path } = input;

  // Build full path
  const fullPath = join(project_path, relative_path);

  // Security: Ensure path is within project
  if (!isPathWithinProject(project_path, fullPath)) {
    logError(
      project_path,
      "read_project_file",
      "PathTraversalAttempt",
      `Path "${relative_path}" resolves to "${fullPath}" which is outside project`,
      { relative_path, resolved: fullPath }
    );
    return {
      success: false,
      error: `Security error: Path "${relative_path}" resolves outside project directory.`,
    };
  }

  logEvent(project_path, "", "read_project_file", "tool_start", "info", {
    relative_path,
    full_path: fullPath,
  });

  try {
    // Check if file exists
    if (!existsSync(fullPath)) {
      return {
        success: false,
        relative_path,
        error: `File not found: ${relative_path}`,
      };
    }

    // Get file stats
    const stats = await stat(fullPath);

    if (stats.isDirectory()) {
      return {
        success: false,
        relative_path,
        error: `Path is a directory, not a file: ${relative_path}`,
      };
    }

    // Read file content
    const content = await readFile(fullPath, "utf-8");

    logEvent(project_path, "", "read_project_file", "tool_end", "info", {
      relative_path,
      size_bytes: stats.size,
      chars: content.length,
    });

    return {
      success: true,
      file_path: fullPath,
      relative_path,
      content,
      size_bytes: stats.size,
    };
  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : "Unknown error";
    logError(
      project_path,
      "read_project_file",
      "ReadError",
      errorMsg,
      { relative_path }
    );

    return {
      success: false,
      relative_path,
      error: `Failed to read file: ${errorMsg}`,
    };
  }
}

// ============================================
// WRITE PROJECT FILE
// ============================================

export const writeProjectFileSchema = z.object({
  project_path: z.string(),
  relative_path: z.string(), // Path relative to project_path
  content: z.string(), // Content to write
  create_dirs: z.boolean().default(true), // Create parent directories if needed
  overwrite: z.boolean().default(true), // Overwrite if file exists
  append: z.boolean().default(false), // Append to end of file instead of replacing
});

export type WriteProjectFileInput = z.infer<typeof writeProjectFileSchema>;

export interface WriteProjectFileResult {
  success: boolean;
  file_path?: string;
  relative_path?: string;
  bytes_written?: number;
  created_dirs?: boolean;
  error?: string;
}

/**
 * Write any file within the project directory
 */
export async function writeProjectFile(
  input: WriteProjectFileInput
): Promise<WriteProjectFileResult> {
  const { project_path, relative_path, content, create_dirs, overwrite, append } = input;

  // Build full path
  const fullPath = join(project_path, relative_path);

  // Security: Ensure path is within project
  if (!isPathWithinProject(project_path, fullPath)) {
    logError(
      project_path,
      "write_project_file",
      "PathTraversalAttempt",
      `Path "${relative_path}" resolves to "${fullPath}" which is outside project`,
      { relative_path, resolved: fullPath }
    );
    return {
      success: false,
      error: `Security error: Path "${relative_path}" resolves outside project directory.`,
    };
  }

  logEvent(project_path, "", "write_project_file", "tool_start", "info", {
    relative_path,
    full_path: fullPath,
    content_length: content.length,
    create_dirs,
    overwrite,
    append,
  });

  try {
    // Check if file exists and overwrite is false
    if (existsSync(fullPath) && !overwrite) {
      return {
        success: false,
        relative_path,
        error: `File already exists: ${relative_path}. Set overwrite=true to replace.`,
      };
    }

    // Create parent directories if needed
    const parentDir = dirname(fullPath);
    let createdDirs = false;

    if (!existsSync(parentDir)) {
      if (create_dirs) {
        await mkdir(parentDir, { recursive: true });
        createdDirs = true;
      } else {
        return {
          success: false,
          relative_path,
          error: `Parent directory does not exist: ${dirname(relative_path)}. Set create_dirs=true to create.`,
        };
      }
    }

    // Determine final content (append mode or replace)
    let finalContent = content;
    const fileExisted = existsSync(fullPath);
    const didAppend = append && fileExisted;

    if (didAppend) {
      const existingContent = await readFile(fullPath, "utf-8");
      finalContent = existingContent + content;
    }

    // Write file
    await writeFile(fullPath, finalContent, "utf-8");

    logEvent(project_path, "", "write_project_file", "tool_end", "info", {
      relative_path,
      bytes_written: Buffer.byteLength(finalContent, "utf-8"),
      appended: didAppend,
      created_dirs: createdDirs,
    });

    return {
      success: true,
      file_path: fullPath,
      relative_path,
      bytes_written: Buffer.byteLength(finalContent, "utf-8"),
      created_dirs: createdDirs,
    };
  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : "Unknown error";
    logError(
      project_path,
      "write_project_file",
      "WriteError",
      errorMsg,
      { relative_path }
    );

    return {
      success: false,
      relative_path,
      error: `Failed to write file: ${errorMsg}`,
    };
  }
}

File: packages/qf-pipeline/src/qf_pipeline/tools/project_files.py
"""
Project file tools for qf-pipeline MCP.

General-purpose file read/write within project directory.
Allows Claude to access ANY file within the project, not just
specific folders like materials/ or questions/.

Security: Only allows access within project_path (no path traversal)

Mirrors the TypeScript implementation in qf-scaffolding.
"""

from pathlib import Path
from typing import Dict, Any, Optional
import os


def is_path_within_project(project_path: str, target_path: str) -> bool:
    """
    Validate that the resolved path is within project_path (prevent path traversal).

    Args:
        project_path: The root project directory
        target_path: The target path to validate

    Returns:
        True if target_path is within project_path, False otherwise
    """
    try:
        # Resolve to absolute paths
        project_resolved = Path(project_path).resolve()
        target_resolved = Path(target_path).resolve()

        # Check if target is relative to project
        target_resolved.relative_to(project_resolved)

        # Also check it's not the same as project_path
        if target_resolved == project_resolved:
            return False

        return True
    except ValueError:
        # relative_to() raises ValueError if not relative
        return False


async def read_project_file(
    project_path: str,
    relative_path: str
) -> Dict[str, Any]:
    """
    Read any file within the project directory.

    Args:
        project_path: Root project directory
        relative_path: Path relative to project_path, e.g. "output/questions.md"

    Returns:
        Dict with success status, content, and metadata
    """
    # Build full path
    full_path = Path(project_path) / relative_path

    # Security: Ensure path is within project
    if not is_path_within_project(project_path, str(full_path)):
        return {
            "success": False,
            "error": f"Security error: Path \"{relative_path}\" resolves outside project directory."
        }

    try:
        # Check if file exists
        if not full_path.exists():
            return {
                "success": False,
                "relative_path": relative_path,
                "error": f"File not found: {relative_path}"
            }

        # Check if it's a directory
        if full_path.is_dir():
            return {
                "success": False,
                "relative_path": relative_path,
                "error": f"Path is a directory, not a file: {relative_path}"
            }

        # Read file content
        content = full_path.read_text(encoding='utf-8')
        stats = full_path.stat()

        return {
            "success": True,
            "file_path": str(full_path),
            "relative_path": relative_path,
            "content": content,
            "size_bytes": stats.st_size
        }

    except Exception as e:
        return {
            "success": False,
            "relative_path": relative_path,
            "error": f"Failed to read file: {str(e)}"
        }


async def write_project_file(
    project_path: str,
    relative_path: str,
    content: str,
    create_dirs: bool = True,
    overwrite: bool = True
) -> Dict[str, Any]:
    """
    Write any file within the project directory.

    Args:
        project_path: Root project directory
        relative_path: Path relative to project_path
        content: Content to write
        create_dirs: Create parent directories if needed (default: True)
        overwrite: Overwrite if file exists (default: True)

    Returns:
        Dict with success status and metadata
    """
    # Build full path
    full_path = Path(project_path) / relative_path

    # Security: Ensure path is within project
    if not is_path_within_project(project_path, str(full_path)):
        return {
            "success": False,
            "error": f"Security error: Path \"{relative_path}\" resolves outside project directory."
        }

    try:
        # Check if file exists and overwrite is False
        if full_path.exists() and not overwrite:
            return {
                "success": False,
                "relative_path": relative_path,
                "error": f"File already exists: {relative_path}. Set overwrite=true to replace."
            }

        # Create parent directories if needed
        parent_dir = full_path.parent
        created_dirs = False

        if not parent_dir.exists():
            if create_dirs:
                parent_dir.mkdir(parents=True, exist_ok=True)
                created_dirs = True
            else:
                return {
                    "success": False,
                    "relative_path": relative_path,
                    "error": f"Parent directory does not exist: {parent_dir.name}. Set create_dirs=true to create."
                }

        # Write file
        full_path.write_text(content, encoding='utf-8')
        bytes_written = len(content.encode('utf-8'))

        return {
            "success": True,
            "file_path": str(full_path),
            "relative_path": relative_path,
            "bytes_written": bytes_written,
            "created_dirs": created_dirs
        }

    except Exception as e:
        return {
            "success": False,
            "relative_path": relative_path,
            "error": f"Failed to write file: {str(e)}"
        }


# Tool registration info for documentation
PROJECT_FILE_TOOLS = [
    {
        "name": "read_project_file",
        "description": "Read any file within a project directory. Security: prevents path traversal outside project.",
        "parameters": {
            "project_path": {"type": "string", "description": "Root project directory"},
            "relative_path": {"type": "string", "description": "Path relative to project_path, e.g. 'output/questions.md'"}
        }
    },
    {
        "name": "write_project_file",
        "description": "Write any file within a project directory. Creates parent dirs by default. Security: prevents path traversal.",
        "parameters": {
            "project_path": {"type": "string", "description": "Root project directory"},
            "relative_path": {"type": "string", "description": "Path relative to project_path"},
            "content": {"type": "string", "description": "Content to write"},
            "create_dirs": {"type": "boolean", "description": "Create parent directories if needed (default: true)", "optional": True},
            "overwrite": {"type": "boolean", "description": "Overwrite if file exists (default: true)", "optional": True}
        }
    }
]
